src/thread.rs

```diff
@@ -0,0 +1,483 @@
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_random_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
+        for (id, thread) in self.threads.iter_enumerated_mut() {
+            if thread.state == ThreadState::Blocked(set) {
+                trace!("unblocking {:?} in blockset {:?}", id, set);
+                thread.state = ThreadState::Enabled;
+                return Some(id);
+            }
+        }
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    fn schedule(&mut self) -> InterpResult<'tcx, SchedulingAction> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
+            for (i, thread) in self.threads.iter_enumerated_mut() {
+                if thread.state == ThreadState::BlockedOnJoin(self.active_thread) {
+                    trace!("unblocking {:?} because {:?} terminated", i, self.active_thread);
+                    thread.state = ThreadState::Enabled;
+                }
+            }
+            return Ok(SchedulingAction::ExecuteDtors);
+        }
+        if self.threads[self.active_thread].state == ThreadState::Enabled
+            && !self.yield_active_thread
+        {
+            return Ok(SchedulingAction::ExecuteStep);
+        }
+        for (id, thread) in self.threads.iter_enumerated() {
+            if thread.state == ThreadState::Enabled {
+                if !(self.yield_active_thread && id == self.active_thread) {
+                    self.active_thread = id;
+                    break;
+                }
+            }
+        }
+        self.yield_active_thread = false;
+        if self.threads[self.active_thread].state == ThreadState::Enabled {
+            return Ok(SchedulingAction::ExecuteStep);
+        }
+        if self.threads.iter().all(|thread| thread.state == ThreadState::Terminated) {
+            Ok(SchedulingAction::Stop)
```

Comment:
As @bjorn3 points out (and I remember the same), the program should actually exit when the main thread stops, not just when all threads stop. I do not know if TLS dtors are run for the remaining threads at that point, we should figure that out.

We should have a test for that.

Reply:
I have updated the scheduler to terminate the program when the main thread terminates. If there are still not terminated threads, the scheduler will `throw_unsup_format` for now.

Reply:
> We should have a test for that.

https://github.com/rust-lang/miri/pull/1284/files#diff-4514f08ca3d99b2ae5b5398344134b25

I have not found a better way to add the error annotation, so suggestions are welcome.

Reply:
Doesn't this mean that detached threads will always make the machine? Or are they fine if they "happen to" be done fast enough?

Reply:
The detached threads are fine if they finish early enough as illustrated by [this test](https://github.com/rust-lang/miri/pull/1284/files#diff-953906240984557ca30301ffd2f18c4cR5-R7):

```rust
fn create_and_detach() {
    thread::spawn(|| ());
}
```

Reply:
That seems rather fragile -- so that test will fail if the scheduler decides to not schedule the new thread? As in, among the many non-deterministic executions that are (in principle) possible, some lead to this "unsupported" error?

My inclination is to
* error with some kind of "leak" message if there are any non-detached threads left
* but error cleanly when there are only detached threads left. we'll just have to figure out if TLS dtors should run or not. (Hopefully not, but some experimentation with real-world code should show that.)

Reply:
> That seems rather fragile -- so that test will fail if the scheduler decides to not schedule the new thread?

> error with some kind of "leak" message if there are any non-detached threads left

So, the idea would be to add an additional leak check similar to this [one](https://github.com/rust-lang/miri/blob/8d9db57a0dee6b2ed013c27fb312cef77a05e323/src/eval.rs#L225)?

> but error cleanly when there are only detached threads left

I assume you meant "**exit** cleanly". I guess that this is not trivial because the memory leak-check will report all memory owned by the detached threads, right?

Also, in my opinion we should have at least an option to report an error in this case because this feels like a scenario where programmers could make mistakes by accident.

> we'll just have to figure out if TLS dtors should run or not. (Hopefully not, but some experimentation with real-world code should show that.)

It seems that at least lib dtors are not executed: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5f24b80ab9ea54dbaf9a795a8561d1b1

Reply:
> So, the idea would be to add an additional leak check similar to this one?

Kind-of, yes.

> I assume you meant "exit cleanly"

*oops* yes I do.

> I guess that this is not trivial because the memory leak-check will report all memory owned by the detached threads, right?

Hm... that is a good point that I had not considered. Okay let's leave that for the future, and error now. "unsupported" is a weird error, but I guess good enough if you also leave a FIXME.